airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatter.java

```diff
@@ -0,0 +1,333 @@
+    // Do not need to iterate through all JSON Object nodes, only first nesting object.
+    if (!fieldsContainRefDefinitionValue.isEmpty()) {
+      fieldsContainRefDefinitionValue.forEach(key -> {
+        if (data.get(key) != null && !data.get(key).isNull()) {
+          data.put(key, data.get(key).toString());
+        }
+      });
+    }
+    addAirbyteColumns(data, recordMessage);
+
+    return data;
+  }
+
+  protected void addAirbyteColumns(ObjectNode data, final AirbyteRecordMessage recordMessage) {
+    final long emittedAtMicroseconds = recordMessage.getEmittedAt();
+    final String formattedEmittedAt = QueryParameterValue.timestamp(emittedAtMicroseconds).getValue();
+
+    data.put(JavaBaseConstants.COLUMN_NAME_AB_ID, UUID.randomUUID().toString());
+    data.put(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, formattedEmittedAt);
+  }
+
+  protected JsonNode formatData(final FieldList fields, final JsonNode root) {
+    // handles empty objects and arrays
+    if (fields == null) {
+      return root;
+    }
+    formatDateTimeFields(fields, root);
+    if (root.isObject()) {
+      return getObjectNode(fields, root);
+    } else if (root.isArray()) {
+      return getArrayNode(fields, root);
+    } else {
+      return root;
+    }
+  }
+
+  protected void formatDateTimeFields(final FieldList fields, final JsonNode root) {
+    final List<String> dateTimeFields = BigQueryUtils.getDateTimeFieldsFromSchema(fields);
+    if (!dateTimeFields.isEmpty() && !root.isNull()) {
+      BigQueryUtils.transformJsonDateTimeToBigDataFormat(dateTimeFields, (ObjectNode) root);
+    }
+  }
+
+  private JsonNode getArrayNode(FieldList fields, JsonNode root) {
+    // Arrays can have only one field
+    final Field arrayField = fields.get(0);
+    // If an array of records, we should use subfields
+    final FieldList subFields;
+    if (arrayField.getSubFields() == null || arrayField.getSubFields().isEmpty()) {
+      subFields = fields;
+    } else {
+      subFields = arrayField.getSubFields();
+    }
+    final JsonNode items = Jsons.jsonNode(MoreIterators.toList(root.elements()).stream()
+        .map(p -> formatData(subFields, p))
+        .collect(Collectors.toList()));
+
+    // "Array of Array of" (nested arrays) are not permitted by BigQuery ("Array of Record of Array of"
+    // is). Turn all "Array of" into "Array of Record of" instead
+    return Jsons.jsonNode(ImmutableMap.of(NESTED_ARRAY_FIELD, items));
+  }
+
+  private JsonNode getObjectNode(FieldList fields, JsonNode root) {
+    final List<String> fieldNames = fields.stream().map(Field::getName).collect(Collectors.toList());
+    return Jsons.jsonNode(Jsons.keys(root).stream()
+        .filter(key -> {
+          final boolean validKey = fieldNames.contains(namingResolver.getIdentifier(key));
+          if (!validKey && !invalidKeys.contains(key)) {
+            LOGGER.warn("Ignoring field {} as it is not defined in catalog", key);
```

Comment:
You are right that even if one field is broken, we will generate at least one additional log record per message. In some cases, it might lead to different issues.
On the other hand, if this happens, something doesn't work properly and should be fixed. Because we lose data.
I propose to collect them here into a set and print the warning on close. We will get one line record in the log + keep the warning visible.